src/main/scala/com/microsoft/hyperspace/index/dataskipping/rule/ApplyDataSkippingIndex.scala

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (2021) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.dataskipping.rule
+
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.Column
+import org.apache.spark.sql.catalyst.expressions.AttributeReference
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, LocalRelation, LogicalPlan}
+import org.apache.spark.sql.execution.datasources.{FileStatusCache, InMemoryFileIndex}
+import org.apache.spark.sql.functions.isnull
+
+import com.microsoft.hyperspace.index.{IndexConstants, IndexLogEntryTags}
+import com.microsoft.hyperspace.index.dataskipping.DataSkippingIndex
+import com.microsoft.hyperspace.index.plans.logical.IndexHadoopFsRelation
+import com.microsoft.hyperspace.index.rules.{ExtractRelation, HyperspaceRule, IndexRankFilter, IndexTypeFilter, QueryPlanIndexFilter}
+import com.microsoft.hyperspace.index.rules.ApplyHyperspace.PlanToSelectedIndexMap
+
+object ApplyDataSkippingIndex extends HyperspaceRule {
+  protected override val filtersOnQueryPlan: Seq[QueryPlanIndexFilter] =
+    IndexTypeFilter[DataSkippingIndex]() :: FilterPlanNodeFilter :: FilterConditionFilter :: Nil
+
+  protected override val indexRanker: IndexRankFilter = FilterRankFilter
+
+  override def applyIndex(plan: LogicalPlan, indexes: PlanToSelectedIndexMap): LogicalPlan = {
+    if (indexes.isEmpty) {
+      return plan
+    }
+    plan match {
+      case filter @ Filter(_, ExtractRelation(relation)) =>
+        val indexLogEntry = indexes(relation.plan)
+        val indexDataPred = new Column(
+          indexLogEntry
+            .getTagValue(plan, IndexLogEntryTags.DATASKIPPING_INDEX_DATA_PREDICATE)
+            .get
+            .getOrElse { return plan })
+        val indexData = spark.read.parquet(indexLogEntry.content.files.map(_.toString): _*)
```

Comment:
Instead of `spark.read.parquet`, could you try to create HadoopFsRelation and try to construct a plan for filteredSourceFilesDf using the relation? I think it's worth trying since we could tag InMemoryFileIndex and avoid "parquet" job for every time

Like this:
https://github.com/microsoft/hyperspace/blob/b60393a9cd92ddabb66e35411b50b11dcb709c02/src/main/scala/com/microsoft/hyperspace/index/covering/CoveringIndexRuleUtils.scala#L113-L115

Reply:
Done